Delete statements will not be created using `clear`

ferrastas opened the pull request:
This is just a demo of an opened issue.

ferrastas committed:
Delete statements will not be created using `clear`.

lcobucci reviewed (dismissed):
Please provide a new test class with its own set of entities instead of modifying an existing one.

ferrastas commented:
Yep, looks like a missing test case of the original fix :P

ostrolucky committed:
Restore ability to clear deferred explicit tracked collections

This was regression from #7862 which tried to respect tracking config
when clearing collections, but this logic can happen in UOW only,
PersistentCollection::clear is triggered too early to know what
is (going to be) persisted.

Fixes #7862

ostrolucky commented:
I've pushed commit with better solution to fix #7758, without this regression

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH7761Test.php:
```diff
@@ -31,6 +31,15 @@ protected function setUp() : void
 
+    protected function tearDown()
+    {
+        $this->_em->clear();
```
This is already handled in the parent, so not needed :+1:

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -380,7 +380,12 @@ public function commit($entity = null)
+                    $this->em->getClassMetadata(get_class($owner = $collectionToDelete->getOwner()))->isChangeTrackingDeferredImplicit() || 
+                    $this->isScheduledForDirtyCheck($owner)) {
+                    $this->getCollectionPersister($collectionToDelete->getMapping())->delete($collectionToDelete);
+                }
```
I think there's a bit too much happening here. We can split things, do an early exit for non-persistent collections, and remove the assignment in condition.

ostrolucky replied:
We can't do early exit, statement must be executed for non-persistent collections. And assignment needs to happen for persistent collections only.

lcobucci replied:
`continue` is an early-exit for the loop :)
I just pushed the changes :+1:

ostrolucky replied:
Yeah I know. I meant we can't do it without duplicating code. If that's fine, then ok

lcobucci reviewed (changes requested):
@ostrolucky thanks a lot for handling it. It's indeed a much better solution. I'll address these two tiny things and we're good to ship this.

lcobucci commented:
@ferrastas @ostrolucky thank you very much :+1:

ferrastas commented:
Thanks to fix it @ostrolucky 🏆
